### PR TITLE
Guid span-based APIs (ctor, TryWriteBytes, TryFormat)

### PR DESCRIPTION
**Pull request by niustephanie**
Implementation of 3 Guid span-based APIs. 
Guid constructor, TryWriteBytes(), and TryFormat() as described in issue #22377.

### CODE REVIEW

File: src/mscorlib/Resources/Strings.resx [hunk 1]
```diff
@@ -293,7 +352,7 @@
     <value>Property Get method was not found.</value>
   </data>
   <data name="Arg_GuidArrayCtor" xml:space="preserve">
-    <value>Byte array for GUID must be exactly {0} bytes long.</value>
+    <value>Constructor for Guid needs exactly {0} bytes.</value>
```
Comment (stephentoub) [inline src/mscorlib/Resources/Strings.resx]: It looks like when you edited this with Visual Studio, it made a few other changes automatically.  Might make sense to revert the changes to this file and then just use notepad to make this one line change.

File: src/mscorlib/src/System/Guid.cs [hunk 2]
```diff
@@ -45,18 +45,22 @@ public struct Guid : IFormattable, IComparable
...
 
         // Creates a new guid from an array of bytes.
-        //
         public Guid(byte[] b)
+            : this(new ReadOnlySpan<byte>(b))
+        {
+        }
+
+        // Creates a new guid from a read-only span.
+        public Guid(ReadOnlySpan<byte> b)
```
Comment (stephentoub) [inline src/mscorlib/src/System/Guid.cs]: It'd be good to verify with some microbenchmarks that the performance of Guid(byte[]) hasn't suffered with this change.  This is how we want to be able to write the code, but as we're still working on span and optimizations related to it, we need to make sure we're not regressing the performance of existing code rewritten to use it.  Hopefully there's no difference and this is fine.  If there is a difference, we'll need to investigate why and what can be done about it.

File: src/mscorlib/src/System/Guid.cs [hunk 3]
```diff
@@ -45,18 +45,22 @@ public struct Guid : IFormattable, IComparable
...
         {
             if (b == null)
                 throw new ArgumentNullException(nameof(b));
```
Comment (stephentoub) [inline src/mscorlib/src/System/Guid.cs]: This check for b == null needs to be done in the Guid(byte[]) ctor, not in the span-based ctor.  I'd suggest removing this if/throw from here, and changing the code above from:
```C#
this(new ReadOnlySpan<byte>(b))
```
to instead be:
```C#
this(new ReadOnlySpan<byte>(b ?? throw new ArgumentNullException(nameof(b)))
```

File: src/mscorlib/src/System/Guid.cs [hunk 4]
```diff
@@ -519,6 +522,13 @@ private static bool TryParseGuid(String g, GuidStyles flags, ref GuidResult resu
         }
 
 
+      /*  private static bool TryParseGuid(String g, GuidStyles flags, ref GuidResult result)
+        {
+            String spanString = new ReadOnlySpan<char>(g, g.Length);
+            return TryParseGuidSpan(spanString, flags, ref result);
+        }*/
```
Comment (stephentoub) [inline src/mscorlib/src/System/Guid.cs]: What is this commented out code?

File: src/mscorlib/src/System/Guid.cs [hunk 5]
```diff
@@ -1014,6 +1024,34 @@ public byte[] ToByteArray()
             return g;
         }
```
Comment (stephentoub) [inline src/mscorlib/src/System/Guid.cs]: From a code reuse perspective, it'd be ideal if this method could just be:
```C#
var g = new byte[16];
TryWriteBytes(g);
return g;
```
but that will add some amount of overhead, so before making such a change, we'd want to understand what if any measurable impact there would be.

File: src/mscorlib/Resources/Strings.resx (hunk 1 above)
Comment (niustephanie) [inline src/mscorlib/Resources/Strings.resx]: I don't recall making this change- is it possible that changes to this file are automatically generated? I'll revert the changes like you suggested.

Comment (niustephanie) [inline src/mscorlib/Resources/Strings.resx]: I don't recall making this change- is it possible that changes to this file are automatically generated? I'll revert the changes like you suggested.

File: src/mscorlib/src/System/Guid.cs (hunk 4 above)
Comment (niustephanie) [inline src/mscorlib/src/System/Guid.cs]: Yep! Will remove

File: src/mscorlib/src/System/Guid.cs (hunk 2 above)
Comment (niustephanie) [inline src/mscorlib/src/System/Guid.cs]: Ok, will do!

File: src/mscorlib/src/System/Guid.cs (hunk 3 above)
Comment (niustephanie) [inline src/mscorlib/src/System/Guid.cs]: Makes sense- thanks for the suggestion.

File: src/mscorlib/src/System/Guid.cs (hunk 4 above)
Comment (niustephanie) [inline src/mscorlib/src/System/Guid.cs]: My bad- meant to delete that

File: src/mscorlib/src/System/Guid.cs [hunk 6]
```diff
@@ -1322,6 +1360,123 @@ unsafe private static int HexsToCharsHexOutput(char* guidChars, int a, int b)
...
+               offset += HexsToChars(guidChars + offset, _a >> 24, _a >> 16);
+               offset += HexsToChars(guidChars + offset, _a >> 8, _a);
+               guidChars[offset++] = ',';
+               guidChars[offset++] = '0';
+               guidChars[offset++] = 'x';
+               offset += HexsToChars(guidChars + offset, _b >> 8, _b);
+               guidChars[offset++] = ',';
+               guidChars[offset++] = '0';
+               guidChars[offset++] = 'x';
+               offset += HexsToChars(guidChars + offset, _c >> 8, _c);
```
Comment (justinvp) [inline src/mscorlib/src/System/Guid.cs]: Nit: It looks like the indentation above is off by a character.

File: src/mscorlib/src/System/Guid.cs [hunk 7]
```diff
@@ -1322,6 +1360,123 @@ unsafe private static int HexsToCharsHexOutput(char* guidChars, int a, int b)
...
+                if (dash) guidChars[offset++] = '-';
+                offset += HexsToChars(guidChars + offset, _c >> 8, _c);
+                if (dash) guidChars[offset++] = '-';
+                offset += HexsToChars(guidChars + offset, _d, _e);
+                if (dash) guidChars[offset++] = '-';
+                offset += HexsToChars(guidChars + offset, _f, _g);
+                offset += HexsToChars(guidChars + offset, _h, _i);
+                offset += HexsToChars(guidChars + offset, _j, _k);
+            }
+    }
```
Comment (justinvp) [inline src/mscorlib/src/System/Guid.cs]: Indentation

File: src/mscorlib/src/System/Guid.cs [hunk 8]
```diff
@@ -1322,6 +1360,123 @@ unsafe private static int HexsToCharsHexOutput(char* guidChars, int a, int b)
...
+            int guidSize;
+
+            if (format == null || format.Length == 0)
+                format = "D";
+
+            if (format.Length != 1)
+            {
+                // all acceptable format strings are of length 1
+                throw new FormatException(SR.Format_InvalidGuidFormatSpecification);
+            }
```
Comment (stephentoub) [inline src/mscorlib/src/System/Guid.cs]: Nit: there should be a blank line after the closing brace

File: src/mscorlib/src/System/Guid.cs [hunk 9]
```diff
@@ -1322,6 +1360,123 @@ unsafe private static int HexsToCharsHexOutput(char* guidChars, int a, int b)
...
+                    FormatGuid(guidChars, ref offset, hex, dash);
+                }
+            }
+
+            charsWritten = guidSize;
+            return true;
+        }
+
         // IFormattable interface
         // We currently ignore provider
```
Comment (stephentoub) [inline src/mscorlib/src/System/Guid.cs]: When I proposed the TryFormat signature with the IFormatProvider argument, it was because Guid.ToString took one... I didn't realize at the time ToString completely ignores it.  @terrajobst, @weshaggard, what do you think, should we remove the IFormatProvider argument from TryFormat?

File: src/mscorlib/src/System/Guid.cs [hunk 10]
```diff
@@ -1322,6 +1360,123 @@ unsafe private static int HexsToCharsHexOutput(char* guidChars, int a, int b)
...
+            {
+                dash = false;
+                guidSize = 32;
+            }
+            else if (formatCh == 'B' || formatCh == 'b')
+            {
+                destination[offset++] = '{';
+                destination[37] = '}';
+                guidSize = 38;
+
```
Comment (stephentoub) [inline src/mscorlib/src/System/Guid.cs]: Nit: unnecessary blank line

File: src/mscorlib/src/System/Guid.cs [hunk 11]
```diff
@@ -1322,6 +1360,123 @@ unsafe private static int HexsToCharsHexOutput(char* guidChars, int a, int b)
...
+                guidSize = 36;
+            }
+            else if (formatCh == 'N' || formatCh == 'n')
+            {
+                dash = false;
+                guidSize = 32;
+            }
+            else if (formatCh == 'B' || formatCh == 'b')
+            {
+                destination[offset++] = '{';
```
Comment (stephentoub) [inline src/mscorlib/src/System/Guid.cs]: If destination isn't large enough, this will throw an out of range exception, but in such a case we need to be returning false.  Same goes for the other places we write into destination.

File: src/mscorlib/src/System/Guid.cs [hunk 12]
```diff
@@ -1322,6 +1360,123 @@ unsafe private static int HexsToCharsHexOutput(char* guidChars, int a, int b)
...
+                destination[37] = '}';
+                guidSize = 38;
+
+            }
+            else if (formatCh == 'P' || formatCh == 'p')
+            {
+                destination[offset++] = '(';
+                destination[37] = ')';
+                guidSize = 38;
+
```
Comment (stephentoub) [inline src/mscorlib/src/System/Guid.cs]: Nit: unnecessary blank line

File: src/mscorlib/src/System/Guid.cs [hunk 13]
```diff
@@ -1322,6 +1360,123 @@ unsafe private static int HexsToCharsHexOutput(char* guidChars, int a, int b)
             return 9;
         }
 
+        // Fills a char* with the appropriate string. 
+        unsafe private void FormatGuid(char* guidChars, ref int offset, bool hex, bool dash)
```
Comment (stephentoub) [inline src/mscorlib/src/System/Guid.cs]: Why is `offset` being passed by ref?

**Comment by jkotas** [comment]
I have submitted #13342 to move Guid.cs to the shared CoreLib partition. It will allow this change to be mirrored to CoreRT automatically.

File: src/mscorlib/src/System/Guid.cs (hunk 13 above)
Comment (niustephanie) [inline src/mscorlib/src/System/Guid.cs]: Realized offset is modified in FormatGuid but isn't used afterwards- changed the pass by ref. Thanks for the catch

File: build-test.cmd [hunk 14]
```diff
@@ -86,9 +86,12 @@ if /i "%1" == "targetsNonWindows"     (set __TargetsWindows=0&set processedArgs=
 if /i "%1" == "Exclude"               (set __Exclude=%2&set processedArgs=!processedArgs! %1 %2&shift&shift&goto Arg_Loop)
 
 if [!processedArgs!]==[] (
-  call set __UnprocessedBuildArgs=!__args!
+  set __UnprocessedBuildArgs=%__args%
```
Comment (JeremyKuhne) [inline build-test.cmd]: It looks like you've pulled in some files that aren't yours. Like you merged rather than rebasing perhaps.

File: src/mscorlib/src/System/Guid.cs [hunk 15]
```diff
@@ -1322,6 +1348,148 @@ unsafe private static int HexsToCharsHexOutput(char* guidChars, int a, int b)
...
+                // all acceptable format strings are of length 1
+                throw new FormatException(SR.Format_InvalidGuidFormatSpecification);
+            }
+
+            char formatCh = format[0];
+            if (formatCh == 'D' || formatCh == 'd')
+            {
+                guidSize = 36;
+            }
+
```
Comment (stephentoub) [inline src/mscorlib/src/System/Guid.cs]: I should have been more clear... my comment above about a blank line after the closing brace was for the standalone if block... no blank line is needed or desired before each else block, though.

File: src/mscorlib/src/System/Guid.cs (hunk 9 above)
Comment (weshaggard) [inline src/mscorlib/src/System/Guid.cs]: I think excluding it makes sense.

Comment (stephentoub) [inline src/mscorlib/src/System/Guid.cs]: Thanks, @weshaggard.  @niustephanie, let's go ahead and remove the provider argument from the new overload.

File: src/mscorlib/src/System/Guid.cs [hunk 16]
```diff
@@ -1322,6 +1348,148 @@ unsafe private static int HexsToCharsHexOutput(char* guidChars, int a, int b)
             return 9;
         }
 
+        // Fills a char* with the appropriate string. 
+        unsafe private void FormatGuid(char* guidChars, int offset, bool hex, bool dash)
```
Comment (danmosemsft) [inline src/mscorlib/src/System/Guid.cs]: I believe we prefer 'private unsafe void ..' ordering.

File: src/mscorlib/src/System/Guid.cs (hunk 9 above)
Comment (niustephanie) [inline src/mscorlib/src/System/Guid.cs]: Got it!

**Comment by niustephanie** [comment]
@dotnet-bot test OSX10.12 x64 Checked Build and Test

**Comment by niustephanie** [comment]
@dotnet-bot test Ubuntu x64 Checked Build and Test
@dotnet-bot test Windows_NT x64 Debug Build and Test

**Comment by niustephanie** [comment]
@dotnet-bot test Windows_NT x64 Release Priority 1 Build and Test
@dotnet-bot test Windows_NT x64 full_opt ryujit CoreCLR Perf Tests Correctness
@dotnet-bot test Windows_NT x64 min_opt ryujit CoreCLR Perf Tests Correctness
@dotnet-bot test Windows_NT x86 Checked Build and Test
@dotnet-bot test Windows_NT x86 full_opt legacy_backend CoreCLR Perf Tests Correctness
@dotnet-bot test Windows_NT x86 full_opt ryujit CoreCLR Perf Tests Correctness
@dotnet-bot test Windows_NT x86 min_opt legacy_backend CoreCLR Perf Tests Correctness
@dotnet-bot test Windows_NT x86 min_opt ryujit CoreCLR Perf Tests Correctness

**Comment by niustephanie** [comment]
@dotnet-bot test OSX10.12 x64 Checked Build and Test

**Comment by niustephanie** [comment]
@dotnet-bot test Windows_NT x64 Debug Build and Test

**Comment by karelz** [comment]
@niustephanie are you still working on it? (just curious)

**Comment by niustephanie** [comment]
@karelz Yes, finishing up perf microbenchmarks to make sure the changes don't have negative impact.

File: src/mscorlib/shared/System/Guid.cs [hunk 17]
```diff
@@ -33,24 +33,63 @@ public partial struct Guid : IFormattable, IComparable, IComparable<Guid>, IEqua
...
+            _c = (short)(b[7] << 8 | b[6]);
+            _d = b[8];
+            _e = b[9];
+            _f = b[10];
+            _g = b[11];
+            _h = b[12];
+            _i = b[13];
+            _j = b[14];
+            _k = b[15];
+        }*/
```
Comment (stephentoub) [inline src/mscorlib/shared/System/Guid.cs]: Why is this commented out rather than deleted?

File: src/mscorlib/shared/System/Guid.cs [hunk 18]
```diff
@@ -33,24 +33,63 @@ public partial struct Guid : IFormattable, IComparable, IComparable<Guid>, IEqua
...
+            _b = (short)(b[5] << 8 | b[4]);
+            _c = (short)(b[7] << 8 | b[6]);
+            _d = b[8];
+            _e = b[9];
+            _f = b[10];
+            _g = b[11];
+            _h = b[12];
+            _i = b[13];
+            _j = b[14];
+            _k = b[15];*/
```
Comment (stephentoub) [inline src/mscorlib/shared/System/Guid.cs]: Should be deleted rather than commented out.

File: src/mscorlib/shared/System/Guid.cs [hunk 19]
```diff
@@ -33,24 +33,63 @@ public partial struct Guid : IFormattable, IComparable, IComparable<Guid>, IEqua
...
 
-
         ////////////////////////////////////////////////////////////////////////////////
         //  Constructors
         ////////////////////////////////////////////////////////////////////////////////
 
         // Creates a new guid from an array of bytes.
-        //
         public Guid(byte[] b)
+        : this(new ReadOnlySpan<byte>(b ?? throw new ArgumentNullException(nameof(b))))
```
Comment (stephentoub) [inline src/mscorlib/shared/System/Guid.cs]: Nit: this is a slightly unusual formatting; often the colon is at the end of the previous line, and then either the `this` is after it on the same line or on the next line but indented.

File: src/mscorlib/shared/System/Guid.cs [hunk 20]
```diff
@@ -235,14 +277,14 @@ internal Exception GetGuidParseException()
...
-        public Guid(String g)
+        public Guid(string g)
         {
             if (g == null)
             {
                 throw new ArgumentNullException(nameof(g));
             }
             Contract.EndContractBlock();
-            this = Guid.Empty;
+            this = Empty;
```
Comment (stephentoub) [inline src/mscorlib/shared/System/Guid.cs]: Nit: it doesn't seem like this line is needed (I realize you didn't add it).

File: src/mscorlib/shared/System/Guid.cs [hunk 21]
```diff
@@ -951,53 +986,68 @@ private static String EatAllWhitespace(String str)
...
         {
-            if (str.Length > i + 1 && str[i] == '0' && (Char.ToLowerInvariant(str[i + 1]) == 'x'))
+            if (str.Length > i + 1 && str[i] == '0' && (char.ToLowerInvariant(str[i + 1]) == 'x'))
                 return true;
             else
                 return false;
         }
 
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public unsafe void WriteByteHelper(Span<byte> destination)
```
Comment (stephentoub) [inline src/mscorlib/shared/System/Guid.cs]: This should be private, not public.

File: src/mscorlib/shared/System/Guid.cs [hunk 22]
```diff
@@ -951,53 +986,68 @@ private static String EatAllWhitespace(String str)
...
+        {
+            fixed (byte* d = &destination.DangerousGetPinnableReference())
```
Comment (stephentoub) [inline src/mscorlib/shared/System/Guid.cs]: Why does this need to be pinned?  Does the perf measurably suffer if you just access destination[x] instead of d[x]?

File: src/mscorlib/shared/System/Guid.cs [hunk 23]
```diff
@@ -1232,80 +1282,77 @@ unsafe private static int HexsToCharsHexOutput(char* guidChars, int a, int b)
...
-
-            if (format.Length != 1)
-            {
-                // all acceptable format strings are of length 1
-                throw new FormatException(SR.Format_InvalidGuidFormatSpecification);
+            bool setBraces = false;
+            bool curly = false;
+            int guidSize;
+
+            switch (char.ToUpperInvariant(format[0]))
```
Comment (stephentoub) [inline src/mscorlib/shared/System/Guid.cs]: Why the change to ToUpperInvariant?  Did this make things faster than just using, e.g.
```C#
case 'D':
case 'd':
    ...;
    break;

case 'N':
case 'n':
   ...
   break;
```
etc.?

File: src/mscorlib/shared/System/Guid.cs [hunk 24]
```diff
@@ -1338,20 +1385,71 @@ public String ToString(String format, IFormatProvider provider)
...
+            if (format == null || format.Length == 0)
+                format = "D";
+
+            // all acceptable format strings are of length 1
+            if (format.Length != 1)
+                throw new FormatException(SR.Format_InvalidGuidFormatSpecification);
+
+            string guidString;
+
+            switch (char.ToUpperInvariant(format[0]))
```
Comment (stephentoub) [inline src/mscorlib/shared/System/Guid.cs]: Same question about ToUpperInvariant.

File: src/mscorlib/shared/System/Guid.cs (hunk 22 above)
Comment (niustephanie) [reply src/mscorlib/shared/System/Guid.cs]: I was testing for perf differences and forgot to revert it- the perf doesn't measurably change. 
Changing to directly access destination[x].

File: src/mscorlib/shared/System/Guid.cs (hunk 23 above)
Comment (niustephanie) [reply src/mscorlib/shared/System/Guid.cs]: This change didn't really impact TryFormat, but it improved ToString somewhat (by about 3.8%). TryParseGuidWithNoStyle also uses ToUpperInvariant in a check (although not in a switch statement), so using it here would be consistent.

Comment (jkotas) [reply src/mscorlib/shared/System/Guid.cs]: This looks suspicious. ToUpperInvariant is relatively expensive operation. switch should be much faster.

File: src/mscorlib/shared/System/Guid.cs [hunk 25]
```diff
@@ -951,53 +947,65 @@ private static String EatAllWhitespace(String str)
...
+                destination[6] = (byte)(_c);
+                destination[7] = (byte)(_c >> 8);
+                destination[8] = _d;
+                destination[9] = _e;
+                destination[10] = _f;
+                destination[11] = _g;
+                destination[12] = _h;
+                destination[13] = _i;
+                destination[14] = _j;
+                destination[15] = _k;
```
Comment (justinvp) [inline src/mscorlib/shared/System/Guid.cs]: Nit: Looks like this is indented too far.

File: src/mscorlib/shared/System/Guid.cs [hunk 26]
```diff
@@ -33,24 +33,25 @@ public partial struct Guid : IFormattable, IComparable, IComparable<Guid>, IEqua
...
 
-
         ////////////////////////////////////////////////////////////////////////////////
         //  Constructors
         ////////////////////////////////////////////////////////////////////////////////
 
         // Creates a new guid from an array of bytes.
-        //
-        public Guid(byte[] b)
+        public Guid(byte[] b):
```
Comment (justinvp) [inline src/mscorlib/shared/System/Guid.cs]: Nit: There should be a space between `)` and `:`

**Comment by niustephanie** [comment]
@dotnet-bot test OSX10.12 x64 Checked Build and Test

**Comment by dezsiszabi** [comment]
Isn't it a general rule with the Try* pattern that these methods are not throwing exceptions (most of the times)?

Taking a quick look at `TryParseExact` and other `TryFormat` methods, those don't throw (I might be wrong here).

This implementation throws when a format string longer than than length 1 is passed in:

    // Returns whether the guid is successfully formatted as a span. 
    public bool TryFormat(Span<char> destination, out int charsWritten, ReadOnlySpan<char> format = default)
    {
        if (format.Length == 0)
            format = "D";

        // all acceptable format strings are of length 1
        if (format.Length != 1) 
            throw new FormatException(SR.Format_InvalidGuidFormatSpecification);

    ......

**Comment by stephentoub** [comment]
> Isn't it a general rule with the Try* pattern that these methods are not throwing exceptions (most of the times)?

No, it's that there's one condition that would otherwise throw and that which now is communicated instead via a Boolean. In this case that condition is whether there's sufficient output space to store the result. Invalid format inputs still result in exceptions.
